### ISSUE TITLE
NPE when using `NavController` from `AppNavModelImpl`

### ISSUE THREAD

**Issue by dan-alcantara-neeva** [neevaco/neeva-android#939]
Versions 0.79.3, 0.80.0, 0.81.0

Stack lines don't make sense, but files do:

https://play.google.com/console/u/0/developers/6544928132232754928/app/4975546713093074924/vitals/crashes/0456d25ff1a0815a76c8fd87278bd8f0/details?installedFrom=PLAY_STORE&days=30&clustering=NEW

```
Exception java.lang.NullPointerException:
  at androidx.navigation.NavController.navigate (NavController.java:3)
  at androidx.navigation.NavController.navigate (NavController.java)
  at com.neeva.app.appnav.AppNavModelImpl.show (AppNavModelImpl.java:9)
  at com.neeva.app.appnav.AppNavModelImpl.showBrowser (AppNavModelImpl.java:2)
  at com.neeva.app.browsing.BaseBrowserWrapper.getUrlBarModel (BaseBrowserWrapper.java)
  at com.neeva.app.NeevaActivity$processIntentInternal$2.invokeSuspend (NeevaActivity.java:2)
  at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith (BaseContinuationImpl.java)
  at kotlinx.coroutines.DispatchedTaskKt.resume (DispatchedTaskKt.java)
  at kotlinx.coroutines.DispatchedTaskKt.resumeUnconfined (DispatchedTaskKt.java:3)
  at kotlinx.coroutines.CancellableContinuationImpl.dispatchResume (CancellableContinuationImpl.java:3)
  at kotlinx.coroutines.CancelledContinuation.makeResumed (CancelledContinuation.java)
  at kotlinx.coroutines.CancellableContinuationImpl.resumeImpl$default (CancellableContinuationImpl.java:3)
  at kotlinx.coroutines.ResumeAwaitOnCompletion.invoke (ResumeAwaitOnCompletion.java)
  at kotlinx.coroutines.JobSupport.completeStateFinalization (JobSupport.java:4)
  at kotlinx.coroutines.JobSupport.tryFinalizeSimpleState (JobSupport.java:3)
  at kotlinx.coroutines.JobSupport.makeCompleting$kotlinx_coroutines_core (JobSupport.java)
  at kotlinx.coroutines.CompletableDeferredImpl.complete (CompletableDeferredImpl.java:4)
  at com.neeva.app.browsing.BaseBrowserWrapper$registerBrowserCallbacks$restorer$4.invoke (BaseBrowserWrapper.java:4)
  at com.neeva.app.browsing.BaseBrowserWrapper$registerBrowserCallbacks$restorer$4.invoke (BaseBrowserWrapper.java:4)
  at com.neeva.app.browsing.BrowserRestoreCallbackImpl.onRestoreCompleted (BrowserRestoreCallbackImpl.java:4)
  at org.chromium.weblayer.Browser$BrowserClientImpl.onRestoreCompleted (Browser.java)
  at org.chromium.weblayer_private.interfaces.IBrowserClient$Stub.onTransact (IBrowserClient.java)
  at android.os.Binder.transact (Binder.java:914)
  at org.chromium.weblayer_private.interfaces.IBrowserClient$Stub$Proxy.onRestoreCompleted (IBrowserClient.java:240)
  at org.chromium.weblayer_private.BrowserImpl.onRestoreCompleted (BrowserImpl.java:555)
  at android.os.MessageQueue.nativePollOnce (MessageQueue.java)
  at android.os.MessageQueue.next (MessageQueue.java:336)
  at android.os.Looper.loop (Looper.java:182)
  at android.app.ActivityThread.main (ActivityThread.java:7563)
  at java.lang.reflect.Method.invoke (Method.java)
  at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run (RuntimeInit.java:539)
  at com.android.internal.os.ZygoteInit.main (ZygoteInit.java:994)
```

**Comment by dan-alcantara-neeva** [comment]
This has jumped up pretty high on the crash list.

**Comment by dan-alcantara-neeva** [comment]
Crash stack doesn't make a whole lot of sense to me; the only way a `NullPointerException` seems like it can be thrown from that location is if `String.toUri()` is passed a null string, which is impossible because of kotlin type safety and other null checks that are performed.

**Comment by dan-alcantara-neeva** [comment]
Forcing the `navController.navigate` call to happen on the IO dispatcher results in a different crash stack:

```
FATAL EXCEPTION: DefaultDispatcher-worker-3
                                                                                                    Process: com.neeva.app.debug, PID: 10221
                                                                                                    java.lang.IllegalStateException: Method setCurrentState must be called on the main thread
                                                                                                    	at androidx.lifecycle.LifecycleRegistry.enforceMainThreadIfNeeded(LifecycleRegistry.java:323)
                                                                                                    	at androidx.lifecycle.LifecycleRegistry.setCurrentState(LifecycleRegistry.java:120)
                                                                                                    	at androidx.navigation.NavBackStackEntry.updateState(NavBackStackEntry.kt:185)
                                                                                                    	at androidx.navigation.NavBackStackEntry.setMaxLifecycle(NavBackStackEntry.kt:156)
                                                                                                    	at androidx.navigation.NavController.updateBackStackLifecycle$navigation_runtime_release(NavController.kt:977)
                                                                                                    	at androidx.navigation.NavController.dispatchOnDestinationChanged(NavController.kt:892)
                                                                                                    	at androidx.navigation.NavController.navigate(NavController.kt:1730)
                                                                                                    	at androidx.navigation.NavController.navigate(NavController.kt:1662)
                                                                                                    	at androidx.navigation.NavController.navigate(NavController.kt:1984)
                                                                                                    	at androidx.navigation.NavController.navigate$default(NavController.kt:1979)
                                                                                                    	at androidx.navigation.NavController.navigate(NavController.kt:1965)
                                                                                                    	at com.neeva.app.appnav.AppNavModelImpl$show$2.invokeSuspend(AppNavModelImpl.kt:109)
                                                                                                    	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
                                                                                                    	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:106)
                                                                                                    	at kotlinx.coroutines.internal.LimitedDispatcher.run(LimitedDispatcher.kt:42)
                                                                                                    	at kotlinx.coroutines.scheduling.TaskImpl.run(Tasks.kt:95)
                                                                                                    	at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:570)
                                                                                                    	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:750)
                                                                                                    	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:677)
                                                                                                    	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:664)
                                                                                                    	Suppressed: kotlinx.coroutines.DiagnosticCoroutineContextException: [StandaloneCoroutine{Cancelling}@e38b1fc, Dispatchers.IO]
```

**Comment by dan-alcantara-neeva** [comment]
Not seeing any crashes for this since my PR landed.

**Comment by dan-alcantara-neeva** [comment]
Eh Play didn't group them together with the other ones -_-

**Comment by dan-alcantara-neeva** [comment]
Looks like the crash stack has changed slightly, so I must have fixed one of the places with my PR at least.  Last time I fixed it by forcing things to run on the UI thread, so I must have missed a pathway.

**Comment by dan-alcantara-neeva** [comment]
```
Exception java.lang.NullPointerException:
  at androidx.navigation.NavController.navigate (NavController.java:3)
  at androidx.navigation.NavController.navigate (NavController.java)
  at com.neeva.app.appnav.AppNavModelImpl.show (AppNavModelImpl.java:9)
  at com.neeva.app.appnav.AppNavModelImpl.showBrowser (AppNavModelImpl.java:2)
  at com.neeva.app.appnav.AppNavModel$DefaultImpls.showBrowser$default (AppNavModel.java:6)
  at com.neeva.app.NeevaActivity$showBrowser$1.invokeSuspend (NeevaActivity.java:6)
  at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith (BaseContinuationImpl.java)
  at kotlinx.coroutines.DispatchedTask.run (DispatchedTask.java)
  at kotlinx.coroutines.EventLoop.processUnconfinedEvent (EventLoop.java:2)
  at kotlinx.coroutines.DispatchedTaskKt.resumeUnconfined (DispatchedTaskKt.java:3)
  at kotlinx.coroutines.CancellableContinuationImpl.dispatchResume (CancellableContinuationImpl.java:3)
  at kotlinx.coroutines.CancelledContinuation.makeResumed (CancelledContinuation.java)
  at kotlinx.coroutines.CancellableContinuationImpl.resumeImpl$default (CancellableContinuationImpl.java:3)
  at kotlinx.coroutines.ResumeAwaitOnCompletion.invoke (ResumeAwaitOnCompletion.java)
  at kotlinx.coroutines.JobSupport.completeStateFinalization (JobSupport.java:4)
  at kotlinx.coroutines.JobSupport.tryFinalizeSimpleState (JobSupport.java:3)
  at kotlinx.coroutines.JobSupport.makeCompleting$kotlinx_coroutines_core (JobSupport.java)
  at kotlinx.coroutines.CompletableDeferredImpl.complete (CompletableDeferredImpl.java:12)
  at com.neeva.app.browsing.BaseBrowserWrapper$registerBrowserCallbacks$restorer$4.invoke (BaseBrowserWrapper.java:12)
  at com.neeva.app.browsing.BaseBrowserWrapper$registerBrowserCallbacks$restorer$4.invoke (BaseBrowserWrapper.java:12)
  at com.neeva.app.browsing.BrowserRestoreCallbackImpl.onRestoreCompleted (BrowserRestoreCallbackImpl.java:4)
  at org.chromium.weblayer.Browser$BrowserClientImpl.onRestoreCompleted (Browser.java)
  at org.chromium.weblayer_private.interfaces.IBrowserClient$Stub.onTransact (IBrowserClient.java)
  at android.os.Binder.transact (Binder.java:1100)
  at org.chromium.weblayer_private.interfaces.IBrowserClient$Stub$Proxy.onRestoreCompleted (IBrowserClient.java:240)
  at org.chromium.weblayer_private.BrowserImpl.onRestoreCompleted (BrowserImpl.java:555)
  at android.os.MessageQueue.nativePollOnce (MessageQueue.java)
  at android.os.MessageQueue.next (MessageQueue.java:335)
  at android.os.Looper.loopOnce (Looper.java:186)
  at android.os.Looper.loop (Looper.java:313)
  at android.app.ActivityThread.main (ActivityThread.java:8663)
  at java.lang.reflect.Method.invoke (Method.java)
  at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run (RuntimeInit.java:567)
  at com.android.internal.os.ZygoteInit.main (ZygoteInit.java:1135)
```

**Comment by Kokobe** [comment]
TLDR of my investigation: 

The crash stack shows that `OnRestoreCompleted()` was called, which makes me think this crash happens when `onCreate` is called. I'm not sure what causes `OnRestoreCompleted()` to be called -- but it seems to be happening at every app startup. 

This [stack overflow ](https://stackoverflow.com/questions/71923628/why-it-throw-an-error-for-navcontroller-as-a-nullpointerexception-in-jetpack-com)post had a NPE while using NavController because one of the destinations it was trying to `navigate()` to was not set up properly (wasn't using `@Composable`). It's likely that `NavController` throws a NPE whenever the view is not set up in a correct way (needs more investigation).

With those 2 pieces of information, here's my theory:
* `WebLayerContainer()` is doing reparenting of its view whenever `onResume()` happens. 
* `OnRestoreCompleted()` happens at least at `onCreate()` but it could happen at other times... this implies that onResume() was eventually called too. 
* As a result of that, the WebLayerContainer() composable is reparented and its reference in `ActivityUI()` is outdated now.
* Now when `showBrowser()` is called, it tries to open the BrowserScaffold() view but it's not configured correctly anymore because ActivityUI didn't recompose or something...

**Comment by Kokobe** [comment]
Going to see if I can reproduce this crash now

**Comment by dan-alcantara-neeva** [comment]
`onRestoreCompleted` is called when Weblayer has finished restoring a profile.  So during startup it'll (normally) load up the regular profile, but it also gets called when you create an incognito profile (by switching over to incognito without any tabs open).

**Comment by Kokobe** [comment]
Tried to comment out reparenting or set the parent to some FrameLayout like this:
```
// in WebLayerContainer()
...
reparentView(it, FrameLayout(context))
```

This didn't cause an NPE 🫤. I think we would need to investigate how NavController throws NPE more. As a workaround, we could just put a try catch around the `navController.navigate()` call in `AppNavModelImpl.kt` since this bug is happening a lot.

**Comment by dan-alcantara-neeva** [comment]
`try / catch` an `RuntimeException` is _always_ a last resort and only something you do if you're sure you can recover safely from it.

**Comment by Kokobe** [comment]
#1239 Fixed it. Not seeing it in the crash logs for version (2910) 0.103.10 🙂
